sigstore-java/src/main/java/dev/sigstore/tuf/Updater.java

```diff
@@ -78,24 +76,32 @@ public static Builder builder() {
   public void update()
       throws IOException, NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException {
     var root = updateRoot();
-    var timestamp = updateTimestamp(root);
-    var snapshot = updateSnapshot(root, timestamp);
-    updateTargets(root, snapshot);
+    // only returns a timestamp value if a more recent timestamp file has been found.
+    var timestampMaybe = updateTimestamp(root);
+    if (timestampMaybe.isPresent()) {
+      var snapshot = updateSnapshot(root, timestampMaybe.get());
+      updateTargets(root, snapshot);
+    }
   }
 
   // https://theupdateframework.github.io/specification/latest/#detailed-client-workflow
   Root updateRoot()
       throws IOException, RoleExpiredException, NoSuchAlgorithmException, InvalidKeySpecException,
-          InvalidKeyException, MetaFileExceedsMaxException, RollbackVersionException,
+          InvalidKeyException, FileExceedsMaxLengthException, RollbackVersionException,
           SignatureVerificationException {
     // 5.3.1) record the time at start and use for expiration checks consistently throughout the
     // update.
     updateStartTime = ZonedDateTime.now(clock);
 
     // 5.3.2) load the trust metadata file (root.json), get version of root.json and the role
     // signature threshold value
-
-    Root trustedRoot = GSON.get().fromJson(Files.readString(trustedRootPath), Root.class);
+    Optional<Root> localRoot = localStore.loadTrustedRoot();
+    Root trustedRoot;
+    if (localRoot.isPresent()) {
+      trustedRoot = localRoot.get();
+    } else {
+      trustedRoot = GSON.get().fromJson(Files.readString(trustedRootPath), Root.class);
+    }
```

Comment:
I like that a lot! If it looks like I haven't coded since Java 7, you'd be right. :)  

Slight issue though, when I move to that approach the IOException from Files.readString becomes a problem for the lambda. Any tricks for cleanly passing that through up the stack?

Reply:
I missed there's a checked exception